### PR TITLE
Issue #12079: Github action "Release perform"

### LINKED ISSUE #12079: Github action "Release perform"
https://github.com/checkstyle/contribution/blob/master/releasenotes-builder/diagrams-manual-trigger.md

we need to make action to do `run Github action "Push perform '10.3.1'"` (action takes human provided version as input, just to make sure it can not be executed unintentionally)
if user provider input is not matching pom - fail build

so we simply make script that do checkout to required tag and do `mvn:release perform`.
this phase suppose to not push anything to github,
but it should able to deploy jars to maven central.

For testing purposes in your test repo you can do release to some folder, you need to configure release plugin to do so.

### PR DESCRIPTION
**Pull request by stoyanK7**
Resolves #12079 

---
I cannot wrap my head around how any of this realease/deploy works and I am stuck. This is my progress until now. I feel like I'm a just doing trial and error hoping something works at this point.

Running
```
 ./.ci/maven-release-perform.sh 10.5.1
```
yields
```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:39 min
[INFO] Finished at: 2022-12-16T19:13:34+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.sonatype.plugins:nexus-staging-maven-plugin:1.6.13:deploy
 (injected-nexus-deploy) on project checkstyle: Failed to deploy artifacts: Could not transfer artifact
 com.puppycrawl.tools:checkstyle:jar:test-sources:10.5.1-20221216.181333-36 
from/to sonatype-nexus-snapshots (https://oss.sonatype.org/content/repositories/snapshots/): 
Transfer failed for https://oss.sonatype.org/content/repositories/snapshots/com/puppycrawl
/tools/checkstyle/10.5.1-SNAPSHOT/checkstyle-10.5.1-20221216.181333-36-test-sources.jar 
401 Unauthorized -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal 
org.sonatype.plugins:nexus-staging-maven-plugin:1.6.13:deploy (injected-nexus-deploy) 
on project checkstyle: Failed to deploy artifacts: Could not transfer artifact 
com.puppycrawl.tools:checkstyle:jar:test-sources:10.5.1-20221216.181333-36 
from/to sonatype-nexus-snapshots (https://oss.sonatype.org/content/repositories/snapshots/):
 Transfer failed for https://oss.sonatype.org/content/repositories/snapshots/com/puppycrawl
/tools/checkstyle/10.5.1-SNAPSHOT/checkstyle-10.5.1-20221216.181333-36-test-sources.jar 401 Unauthorized

```

### CODE REVIEW

**Comment by romani** [comment]
Please never hesitate to ask ask questions. Draft PR with questions is good form. I will reply soon

File: .ci/maven-release-perform.sh [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+SKIP_TEST="-DskipTests -DskipITs"
+SKIP_CHECKSTYLE="-Dcheckstyle.ant.skip=true -Dcheckstyle.skip=true"
+SKIP_OTHERS="-Dpmd.skip=true -Dspotbugs.skip=true -Djacoco.skip=true -Dxml.skip=true"
+
+git checkout "checkstyle-$TARGET_VERSION"
+echo "Deploying jars to maven central (release:perform) ..."
+#mvn -e --no-transfer-progress release:perform -Pno-validations -DconnectionUrl=scm:git:git@github.com:stoyanK7/checkstyle.git -Dtag=checkstyle-"$TARGET_VERSION" -Dcs.maven.directory=/home/stoyan/test
+#mvn -e --no-transfer-progress -Pgpg release:perform -Pno-validations -DconnectionUrl=scm:git:git@github.com:stoyanK7/checkstyle.git -Dtag=checkstyle-"$TARGET_VERSION" -Dcs.maven.directory=/home/stoyan/test \
+#  -Darguments="$SKIP_TEST $SKIP_CHECKSTYLE $SKIP_OTHERS"
+mvn clean package -Passembly deploy -e -Dcs.maven.directory=/home/stoyan/test
```
Comment (romani) [inline .ci/maven-release-perform.sh:32]: we simply need to run `-Pgpg release:perform`
https://github.com/checkstyle/checkstyle/blob/2a3d28489f8bf52cc55636d43926d99bada7fa11/release.sh#L41-L43

`-Pgpg` is profile to from settings.xml to take all secrets to have access to to sonatype maven central for jars upload.

`release:perform` is maven command to generate jars and upload to maven central. Maven central is pre-configured by parent pom (as you can see it is parent for all artifacts that want to be deployed to maven central)
https://github.com/checkstyle/checkstyle/blob/2a3d28489f8bf52cc55636d43926d99bada7fa11/pom.xml#L6-L11

**Review by romani — CHANGES_REQUESTED** [review]
items:

**Comment by romani** [comment]
https://github.com/checkstyle/checkstyle/blob/master/.ci/prepare-settings.sh is script that generates settings.xml file with content from secrets.

File: .ci/maven-release-perform.sh (hunk 1 above)
Comment (stoyanK7) [reply .ci/maven-release-perform.sh]: Should be done.

**Comment by stoyanK7** [comment]
Waiting for review..

File: .ci/maven-release-perform.sh (hunk 1 above)
Comment (stoyanK7) [reply .ci/maven-release-perform.sh]: @romani ping

**Comment by romani** [comment]
@stoyanK7 , please make PR ready for review, code is is good enough, polish it, and lets merge.
if you have problems to test it, please share it, I might help or merge to test in real release.

**Comment by stoyanK7** [comment]
@romani Done. Successfully deployed to a local folder with this commit: https://github.com/stoyanK7/checkstyle/commit/3a6d340dd01b6a2c98aac8aa29dd688b6c2be708

```sh
    [INFO] --- maven-deploy-plugin:3.0.0:deploy (default-deploy) @ checkstyle ---
    [INFO] Using alternate deployment repository internal.repo::file:///home/stoyan
    Downloading from internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/maven-metadata.xml
    Uploading to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1.pom
    Uploaded to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1.pom (224 kB at 9.7 MB/s)
    Uploading to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1.jar
    Uploaded to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1.jar (2.0 MB at 34 MB/s)
    Uploading to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1-tests.jar
    Uploaded to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1-tests.jar (3.7 MB at 76 MB/s)
    Uploading to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1-test-sources.jar
    Uploaded to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1-test-sources.jar (3.8 MB at 78 MB/s)
    Uploading to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1-sources.jar
    Uploaded to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1-sources.jar (1.8 MB at 80 MB/s)
    Uploading to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1-sources.jar
    Uploaded to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1-sources.jar (1.8 MB at 80 MB/s)
    Uploading to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1-javadoc.jar
    Uploaded to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1-javadoc.jar (7.8 MB at 86 MB/s)
    Uploading to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1.jar.asc
    Uploaded to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1.jar.asc (659 B at 659 kB/s)
    Uploading to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1.pom.asc
    Uploaded to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1.pom.asc (659 B)
    Uploading to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1-tests.jar.asc
    Uploaded to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1-tests.jar.asc (659 B)
    Uploading to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1-test-sources.jar.asc
    Uploaded to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1-test-sources.jar.asc (659 B)
    Uploading to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1-sources.jar.asc
    Uploaded to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1-sources.jar.asc (659 B)
    Uploading to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1-sources.jar.asc
    Uploaded to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1-sources.jar.asc (659 B at 659 kB/s)
    Uploading to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1-javadoc.jar.asc
    Uploaded to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/checkstyle-10.7.1-20230203.101110-1-javadoc.jar.asc (659 B)
    Downloading from internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/maven-metadata.xml
    Uploading to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/maven-metadata.xml
    Uploaded to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/10.7.1-SNAPSHOT/maven-metadata.xml (2.9 kB)
    Uploading to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/maven-metadata.xml
    Uploaded to internal.repo: file:///home/stoyan/com/puppycrawl/tools/checkstyle/maven-metadata.xml (291 B)
    [INFO] ------------------------------------------------------------------------
    [INFO] BUILD SUCCESS
    [INFO] ------------------------------------------------------------------------
    [INFO] Total time:  49.109 s
    [INFO] Finished at: 2023-02-03T11:11:10+01:00
    [INFO] ------------------------------------------------------------------------
[INFO] Cleaning up after release...
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:07 min
[INFO] Finished at: 2023-02-03T11:11:10+01:00
[INFO] ------------------------------------------------------------------------

```

Screenshot from 2023-02-03 11-12-37

**Review by romani — APPROVED** [review]
ok to merge, we will test in real in few day, we will try to do patch release before end of month

**Pull request merged**